staging/src/k8s.io/apiserver/pkg/cel/library/ip_addrs.go

```diff
@@ -0,0 +1,199 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package library
+
+import (
+	"fmt"
+	"net/netip"
+
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+
+	apiservercel "k8s.io/apiserver/pkg/cel"
+)
+
+// IPAddr provides a CEL function library extension of IP address parsing functions.
+//
+// ipAddr
+//
+// Converts a string to an IP address or results in an error if the string is not a valid IP address.
+// The IP address must be an IPv4 or IPv6 address.
+// IPv4-mapped IPv6 addresses (e.g. ::ffff:1.2.3.4) are not allowed.
+// IP addresses with zones (e.g. fe80::1%eth0) are not allowed.
+//
+//	ipAddr(<string>) <IPAddr>
+//
+// Examples:
+//
+//	ipAddr('127.0.0.1') // returns an IPv4 address
```

Comment:
I also think we should not allow it ... if this new validator is added to an existing type, I'd lean on ratcheting validation to deal with letting existing data be updated

Reply:
The consensus a year and a half ago in #108074 was that we should stop accepting invalid (leading 0) values for all newly-created objects / newly-updated fields of existing objects, even where that is technically an API break, because security. I'm pretty sure the only reason we never actually implemented it is because the validation code as currently written makes it hard to do that without also forcing stricter revalidation of *not*-updated fields of existing objects and breaking the world.

Reply:
Agree with Dan and Jordan, let's not carry forward the sloppy parsers

Reply:
Since I'm using `netip` to parse the addresses in both the ip and cidr implementations, the library doesn't allow leading zeros, so I just need to update the documentation to reflect that.
An example error from the library:
```
ParseAddr("10.0.0.010"): IPv4 field has octet with leading zero
```